### ISSUE TITLE
Support for Protobuf V2 API

### ISSUE THREAD

**Issue by tpaschalis** [beatlabs/patron#168]
## Is your feature request related to a problem? Please describe

A new Go API for Protocol Buffers [was released](https://blog.golang.org/a-new-go-api-for-protocol-buffers), which Patron could add support for.

## Describe the solution
Support would initially be included in the generic `encoding` package.   

Afterwards, changes in various clients (eg. in the custom Kafka encoders/decoders or header encoding in sync/http) can be gradually implemented.

## Additional context
None, for now.

**Comment by mantzas** [comment]
Let's wait for this one before we jump into it.

**Comment by mantzas** [comment]
@tpaschalis @drakos74 I believe we should upgrade this one in order to be able to use the much newer api...
Should we include this to v1?

**Comment by tpaschalis** [comment]
I'm not certain how we should proceed with this to be honest.

The v2 Protobuf API is not backwards compatible with v1. They're completely separate packages with different import paths (`github.com/golang/protobuf`  for v1, `google.golang.org/protobuf` for v2) and have different opinions on what a `proto.Message` should be. I'm not even sure if code using the v1 protobuf message would work with the v2 API. 

A series of steps could be
1) Understand the documentation
2) Try to run the examples using the old definitions but upgrading the Protobuf API
3) Reach out to end users and see how more complex code works , in real-world scenarios.

**Comment by mantzas** [comment]
@tpaschalis i have read that they will keep both of the packages indefinitely which is very odd and we will face packages including both versions... what a mess...

**Comment by tpaschalis** [comment]
**TL;DR** I think that we could upgrade to `protobuf@v1.4.0` so we get the latest APIv1 and pave the road for APIv2 at the same time by letting users make their code v2-compatible gradually. What do you think?

@mantzas a mess indeed.. and it's not even a 'true' v2, but `v1.20`, supposedly since they don't expect the v1 to reach that version ever.  It's the umpteenth time I've read https://blog.golang.org/protobuf-apiv2. So in short we have
```
github.com/golang/protobuf@v1.3.4     API v1

github.com/golang/protobuf@v1.4.0     APIv1 implemented by APIv2; contains helpers for converting between 
                                      APIv1 and APIv2 proto.Message interfaces and ease the transition

google.golang.org/protobuf@v1.20.0    API V2, depends on v1.4.0
````

*Dependency management : such a hard task that even Google has given up*
